heaps/min_heap.py

```diff
@@ -57,18 +68,43 @@ def heap_up(self, index):
             property is reestablished.
             
             This could be **very** helpful for the add method.
-            Time complexity: ?
-            Space complexity: ?
+            Time complexity: O(log n)
+            Space complexity: O(log n)
         """
-        pass
+        if index == 0:
+            return
+
+        if index % 2 == 0:
+            # right node parent
+            parent_index = int((index - 2) / 2)
+        else:
+            # left node parent
+            parent_index = int((index - 1) / 2)
+
+        if(parent_index < 0):
+            return
+
+        if self.store[parent_index].key > self.store[index].key:
+            self.swap(parent_index, index)
+            self.heap_up(parent_index)
 
     def heap_down(self, index):
         """ This helper method takes an index and 
             moves the corresponding element down the heap if it's 
             larger than either of its children and continues until
             the heap property is reestablished.
         """
-        pass
+        left_child = index * 2 + 1
+        right_child = index * 2 + 2
+
+        if len(self.store) > (left_child) and self.store[index].key > self.store[left_child].key:
+            self.swap(index, (left_child))
+            self.heap_down((left_child))
+            self.heap_down(index)
+        elif len(self.store) > (right_child) and self.store[index].key > self.store[right_child].key:
+            self.swap(index, (right_child))
+            self.heap_down((right_child))
+            self.heap_down(index)
```

Comment:
You can reduce your recursive calls to `heap_down` here. Instead of choosing to swap with the left child automatically if it exists, choose to swap `index` with whichever child is smaller. 

If you swap with the smaller of the two children, that child (as the new parent of `index` and the other child) will automatically be smaller than both of it's new children and you won't have to call `heap_down` on it again.